pkg/tun/tun_darwin.go

```diff
@@ -0,0 +1,246 @@
+package tun
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+	"unsafe"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+	"golang.org/x/sys/unix"
+
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/buffer"
+)
+
+const sysProtoControl = 2
+const uTunOptIfName = 2
+const uTunControlName = "com.apple.net.utun_control"
+
+type Device struct {
+	*os.File
+	name string
+}
+
+// OpenTun creates a new TUN device and ensures that it is up and running.
+func OpenTun() (*Device, error) {
+	fd, err := unix.Socket(unix.AF_SYSTEM, unix.SOCK_DGRAM, sysProtoControl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open DGRAM socket: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			_ = unix.Close(fd)
+		}
+	}()
+
+	info := &unix.CtlInfo{}
+	copy(info.Name[:], uTunControlName)
+	if err = unix.IoctlCtlInfo(fd, info); err != nil {
+		return nil, fmt.Errorf("failed to getBuffer IOCTL info for %s: %w", uTunControlName, err)
+	}
+
+	if err = unix.Connect(fd, &unix.SockaddrCtl{ID: info.Id, Unit: 0}); err != nil {
+		return nil, err
+	}
+
+	if err = unix.SetNonblock(fd, true); err != nil {
+		return nil, err
+	}
+
+	name, err := unix.GetsockoptString(fd, sysProtoControl, uTunOptIfName)
+	if err != nil {
+		return nil, err
+	}
+	return &Device{File: os.NewFile(uintptr(fd), ""), name: name}, nil
+}
+
+// AddSubnet adds a subnet to this TUN device and creates a route for that subnet which
+// is associated with the device (removing the device will automatically remove the route).
+func (t *Device) AddSubnet(_ context.Context, subnet *net.IPNet) error {
```

Comment:
A common idiom (at least in the stdlib) is to declare `t.Method()` in a common file, and have it just call `t.method()` which is defined in in the platform-specific file.  That way the godoc only needs to be written once.  But that probably matters more in libraries, and not so much in applications.